UD-5975 Add nights duration to daterangepicker

ssuttner opened the pull request:
<!-- Description -->

JIRA: [UD-5975](https://urbandoor.atlassian.net/browse/UD-5975)

<!--
## Security

Does this PR have security implications (manipulates user input in anyway, 
interacts with sessions, introduces a new gem, a new library, etc)? If so, 
how did you address it? If not sure, please specify so a reviewer can assess it. 
-->

* Any kind of feedback is highly appreciated.
* I'm still learning how things should be approached and taken care of in this repo.
* Test are yet to be implemented.


<!--
## How to Review

* Commit by commit recommended.
-->

## Screenshots

* Before
<img width="502" alt="screen shot 2018-09-10 at 12 12 45 pm" src="https://user-images.githubusercontent.com/998876/45306370-e1e88f00-b4f2-11e8-9b11-fe0a57810bb6.png">

* After
<img width="517" alt="screen shot 2018-09-10 at 11 57 29 am" src="https://user-images.githubusercontent.com/998876/45306373-e3b25280-b4f2-11e8-9f1d-bbc03b41bddf.png">





<!--
Final checklist:

* [ ] Read through the entire diff again
* [ ] Tag applicable labels
* [ ] Add specific reviewers
* [ ] Comment pre-emptively on dubious lines
-->

ianwremmel commented on README.md:
```diff
@@ -355,6 +364,32 @@ most everything in `./features/support` are taken pretty much directly from the
 [WDIO Cucumber Boilerplate](https://github.com/webdriverio/cucumber-boilerplate)
 (with adjustments made to support CommonJS instead of ESM).
 
+Simly run:
```
Please remove this section; I've got a PR that's going to include this in a new contributing.md

ianwremmel commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -1,20 +1,46 @@
-    data-press-daterangepicker-input
-  >
+  <div :class="[label !== 'null' ? 'field' : '']">
+    <label
```
`<label>`s are for describing input fields. This should really just be a `<p>` or a `<div>`.

ssuttner replied:
@ianwremmel this is actually a `<label>` on purpose.
I need a label for this field on the containing `form`.

Before PRESS

```
label :start_date + input :start_date
label :end_date + input :end_date
```

After PRESS
```
no-label + drp
```

ianwremmel replied:
Oh, I see, I misread and thought this was containing the duration value

ianwremmel commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -1,20 +1,46 @@
-    :value="localeStringeFromValue"
-    data-press-daterangepicker-input
-  >
+  <div :class="[label !== 'null' ? 'field' : '']">
```
I think the class binding will be a lot cleaner as objects instead of arrays:

```vue
<div :class="{field: label !== 'null'}">
```

ianwremmel commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -1,20 +1,46 @@
+    >
+      {{ label }}
+    </label>
+    <div :class="['ui', displayDuration ? 'right labeled' : '', 'input']">
```
vue does a nice job of merging the `class` and `v-bind:class` directives, so maybe move the constant classes to `class`.

```vue
<div class="ui input" :class="{right: !!displayDuration, labeled: !!displayDuration}">

SashaBayan commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -64,12 +93,27 @@ export default {
         return this.value[this.endKey];
       }
     },
+    duration() {
```
`duration` seems a bit ambiguous here -- perhaps `stayLength` is a better term?

ssuttner replied:
Updating!

SashaBayan commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -64,12 +93,27 @@ export default {
       }
+    },
+    durationString() {
+      if (this.duration == 1) {
```
prefer `===`

SashaBayan commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -64,12 +93,27 @@ export default {
         return this.value[this.endKey];
       }
     },
+    duration() {
```
@ianwremmel -- is it possible to add return types to these methods?

SashaBayan commented:
Having 3 separate DRP concerns me (one w the stay length duration, the other with a label, and the other one "unnested"). 

Our Vue DRP should be able to expose its internal state such that external components or a global state machine can consume and store data for other components to use. This kind of architecture would allow us to have a separate text field that display the nights duration. Having the nights duration component tagged onto the DRP seems much more like an anti-semantic anti-pattern. I am not sure how much flexibility we're afforded for data exposure (since our javascripless ERB template cannot dip into the DRP and extract its internal state), so we may have to live with this way of doing things.

SashaBayan commented:
Can we re-use the "unnested" DRP in the other files such so we can be more DRY?

SashaBayan commented:
Sorry -- I may have misread our test code display files as the component files. Ignore the bloating comment above

SashaBayan commented on src/components/daterangepicker/daterangepicker.vue:
```diff
@@ -1,20 +1,49 @@
+      class="ui input"
+    >
+      <input
+        :value="localeStringeFromValue"
```
`Stringe` -- spelling error

ssuttner replied:
Fixing!

ssuttner commented:
@ianwremmel I just rebased and added compatibility with the existing `formless` version used on the home page.

<img width="1434" alt="screen shot 2018-09-26 at 4 16 38 pm" src="https://user-images.githubusercontent.com/998876/46111159-4cb8ec00-c1bc-11e8-8112-3f210ed4c0d5.png">
<img width="720" alt="screen shot 2018-09-26 at 4 16 58 pm" src="https://user-images.githubusercontent.com/998876/46111158-4cb8ec00-c1bc-11e8-9b75-a9386b87d8db.png">



@SashaBayan here is how the `server` code looks like:
```
<div data-press-component="daterangepicker" data-press-daterangepicker-display-stay-length data-press-daterangepicker-label="Stay Range">
  <div class="two fields">
    <div class="field">
      <%= f.input :start_date, label: t('pages.housing_request_form.start_date_label'), html5: true %>
    </div>
    <div class="field">
      <%= f.input :end_date, label: t('pages.housing_request_form.end_date_label'), html5: true %>
    </div>
  </div>
</div>
``` 
I like it because it doesn't affect the visuals of the form in case press doesn't kick in. This is a comment in relation to [this discussion thread](https://github.com/UrbanDoor/server/pull/7241).

I think support for invalid (red) from fields is still needed (as I explained to @SashaBayan via DM) but I'm not sure how much of that is part of the scope of this `stay-length` story.

@ianwremmel I know your component story is also about to get merged, I'm not sure how you'd like to proceed in terms of that.

SashaBayan commented on src/components/daterangepicker/index.js:
```diff
@@ -54,10 +54,33 @@ export default class DateRangePicker extends PressComponentBase {
+    const endField = endEl.closest('.field');
+    drp.setAttribute(
+      'v-bind:validation-error',
+      (startField && startField.classList.contains('error')) ||
```
we should extract this into a computed or variable for clarity. something like:
```
const hasErrors = (startField && startField.classList.contains('error')) ||(endField && endField.classList.contains('error'))
drp.setAttribute(
  'v-bind:validation-error',
   hasErrors
)
```
